Make swraid optional

lheckemann opened the pull request:
###### Description of changes

This is a draft based on [some discussion](https://matrix.to/#/!PSmBFWNKoXmlQBzUQf:helsinki-systems.de/$BCknQ22raVC39s3l3GJz_9y0-VGj38CLtmUGfBR5TqY?via=nixos.org&via=matrix.org&via=helsinki-systems.de) in the Matrix room for systemd stage1.

cc @ElvishJerricco @ajs124 

###### Things done

- [x] Run swraid installer tests with both scripted and systemd stage-1
- [x] Write release notes (don't rely on people to use stateVersion correctly...)
- [x] ~~Apply the same to bcache too~~ let's do this in a later PR
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lheckemann committed:
nixos/swraid: make entire module optional

swraid support will now only be enabled by default if stateVersion is
older than 23.11. nixos-generate-config will now generate explicit
config for enabling support if needed.

ElvishJerricco commented on nixos/modules/tasks/swraid.nix:
```diff
@@ -6,7 +6,7 @@ in {
   options.boot.initrd.services.swraid = {
     enable = (lib.mkEnableOption "swraid support using mdadm") // {
-      visible = false; # only has effect when the new stage 1 is in place
+      default = lib.versionOlder config.system.stateVersion "22.11";
```
Maybe change defaultText as well? Just to make it clear.

ElvishJerricco commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -514,6 +514,13 @@ sub in {
         }
+        if ($dmUuid =~ /^LVM-/)
+        {
+            push @attrs, "  boot.initrd.services.lvm.enable = true;";
```
It was mentioned on matrix that this might result in the line being added to the config multiple times if there are multiple LVM/md devices. If that's the case, we definitely need to guard against it.

ElvishJerricco reviewed (changes requested):
We should probably go ahead and do the same for LVM in this PR.

ElvishJerricco commented on nixos/modules/profiles/installation-device.nix:
```diff
@@ -105,6 +105,11 @@ with lib;
         systemdStage1Network
       ];
 
+    boot.initrd.services = {
+      lvm.enable = true;
+      swraid.enable = true;
+    };
+
```
Why is this necessary?

lheckemann replied:
In order to provide `mdadm` and the LVM tools in installers.

ElvishJerricco replied:
@lheckemann Not that it matters much, but shouldn't those just be added to `systemPackages`? Shouldn't need to be in initrd for the installer since the installer doesn't boot off LVM or mdraid.

lheckemann replied:
I'm not sure about LVM and swraid actually, but I know that cryptsetup fails in weird ways (opening a volume just hangs) if you only install the package and not the udev rules. The installer should have full support, not just the packages available.

ElvishJerricco replied:
Oh sure. I guess I'm just pointing out that this does *more* than that by putting it in initrd as well, but again, it really doesn't matter much.

ElvishJerricco replied:
Actually wait, isn't that a huge problem if lvm and swraid rules aren't in stage *2* by default anymore? We only want to make them optional in stage 1; stage 2 should have them and the ISO shouldn't have to enable them itself.

lheckemann replied:
They should not be in stage 2 either unless specifically requested by the user. The current behaviour, which places LVM in _all_ non-container system closures and activates all LVM volume groups without the user having configured anything is surprising at best and harmful at worst.

The `nixos-generate-config` changes, release notes, and `stateVersion`-based default should prevent this from breaking without notice for people who have installed their systems on LVM or mdraid.

ElvishJerricco commented:
Oh, looks like we ought to do this for [bcache](https://github.com/NixOS/nixpkgs/blob/88199c6df95b3f7a8b8bbcaf4dd8977612f219c0/nixos/modules/tasks/bcache.nix#L18) as well.

aktaboot commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -514,8 +516,21 @@ sub in {
+    push @attrs, "boot.initrd.services.lvm.enable = true;\n\n";
+}
+if ($useSwraid) {
+    push @attrs, "boot.initrd.services.swraid.enable = true;\n\n"
```
is this missing a `;`

lheckemann commented:
I've rebased and switched the release notes and stateVersion condition to 23.11.

I think that the options should be moved out of `boot.initrd.services.swraid` and to something like `boot.swraid`, since this config is relevant to stage-2 as well. I'd start with enabling it in stage-1 iff it's enabled for stage-2, and not worry about distinguishing between stage-2-only and both-stages support for now.

I don't think we need to provide backwards-compatibility aliases since stage-1 systemd is still marked as experimental.

Thoughts?

lheckemann committed:
boot.initrd.services.swraid -> boot.swraid

Since the option affects both stage-1 and stage-2, it does not make
sense to keep it within the boot.initrd namespace.

lheckemann committed:
stage-1: Only copy mdadm and mdmon into initramfs if enabled

ElvishJerricco reviewed (approved):
This looks good to me, but I guess this means we're going to do the same changes for LVM and bcache in separate PRs?

lheckemann commented:
Yes. This was enough of a fuss that I wanted to keep it separate from the LVM and bcache changes.

SuperSandro2000 commented on nixos/modules/tasks/swraid.nix:
```diff
@@ -1,47 +1,71 @@
 in {
 
-  options.boot.initrd.services.swraid = {
+  options.boot.swraid = {
```
We missed adding a mkRenamedOption for this

```
       error: attribute 'swraid' missing

       at /nix/store/6vrnicwhzppfy58v8l8jhpjplzckd74k-source/configuration.nix:103:78:

          102|       # https://github.com/numtide/srvos/blob/master/common/default.nix#L20
          103|       systemd.enable = lib.mkDefault (!config.boot.initrd.network.enable && !config.boot.initrd.services.swraid.enable && !config.boot.isContainer);
             |                                                                              ^
          104|     };
```

yayayayaka replied:
Opened #243577
